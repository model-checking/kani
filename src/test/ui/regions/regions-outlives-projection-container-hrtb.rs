// Test that structs with higher-ranked where clauses don't generate
// "outlives" requirements. Issue #22246.

// revisions: migrate nll
//[nll]compile-flags: -Z borrowck=mir

// Since we are testing nll (and migration) explicitly as a separate
// revisions, don't worry about the --compare-mode=nll on this test.

// ignore-compare-mode-nll

#![allow(dead_code)]

pub trait TheTrait<'b> {
    type TheAssocType;
}

pub struct TheType<'b> {
    m: [fn(&'b()); 0]
}

impl<'a,'b> TheTrait<'a> for TheType<'b> {
    type TheAssocType = &'b ();
}

pub struct WithHrAssoc<T>
    where for<'a> T : TheTrait<'a>
{
    m: [T; 0]
}

fn with_assoc<'a,'b>() {
    // We get an error because 'b:'a does not hold:

    let _: &'a WithHrAssoc<TheType<'b>> = loop { };
    //[migrate]~^ ERROR reference has a longer lifetime
    //[nll]~^^ ERROR lifetime may not live long enough
}

pub trait TheSubTrait : for<'a> TheTrait<'a> {
}

impl<'b> TheSubTrait for TheType<'b> { }

pub struct WithHrAssocSub<T>
    where T : TheSubTrait
{
    m: [T; 0]
}

fn with_assoc_sub<'a,'b>() {
    // The error here is just because `'b:'a` must hold for the type
    // below to be well-formed, it is not related to the HR relation.

    let _: &'a WithHrAssocSub<TheType<'b>> = loop { };
    //[migrate]~^ ERROR reference has a longer lifetime
    //[nll]~^^ ERROR lifetime may not live long enough
}


fn main() {
}
