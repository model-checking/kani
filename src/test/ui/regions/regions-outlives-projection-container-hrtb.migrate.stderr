error[E0491]: in type `&'a WithHrAssoc<TheType<'b>>`, reference has a longer lifetime than the data it references
  --> $DIR/regions-outlives-projection-container-hrtb.rs:35:12
   |
LL |     let _: &'a WithHrAssoc<TheType<'b>> = loop { };
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: the pointer is valid for the lifetime `'a` as defined here
  --> $DIR/regions-outlives-projection-container-hrtb.rs:32:15
   |
LL | fn with_assoc<'a,'b>() {
   |               ^^
note: but the referenced data is only valid for the lifetime `'b` as defined here
  --> $DIR/regions-outlives-projection-container-hrtb.rs:32:18
   |
LL | fn with_assoc<'a,'b>() {
   |                  ^^

error[E0491]: in type `&'a WithHrAssocSub<TheType<'b>>`, reference has a longer lifetime than the data it references
  --> $DIR/regions-outlives-projection-container-hrtb.rs:55:12
   |
LL |     let _: &'a WithHrAssocSub<TheType<'b>> = loop { };
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: the pointer is valid for the lifetime `'a` as defined here
  --> $DIR/regions-outlives-projection-container-hrtb.rs:51:19
   |
LL | fn with_assoc_sub<'a,'b>() {
   |                   ^^
note: but the referenced data is only valid for the lifetime `'b` as defined here
  --> $DIR/regions-outlives-projection-container-hrtb.rs:51:22
   |
LL | fn with_assoc_sub<'a,'b>() {
   |                      ^^

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0491`.
